allow org members to view circle members pages

levity opened the pull request:
This PR makes the circle members page accessible to org members. This means:

- As an org member (not a circle member), you can mouse over circles in the org page and see a link to the Members page, which you can click on
- When you do, the Members page will be loaded, and the side nav will show the circle you've selected (which doesn't show otherwise, since you're not a member
- The Members page has limited content: there's no admin button or Leave Circle button, of course

### Testing

Since there is no way in the UI to add a member directly to an org yet, you can do this from the REPL: run `await createMembersFromCircles(orgId)`, where `orgId` is the org you want to add members to. All circle members in that org will be made org members as well.

If you want to scrutinize the permissions setup, you can remove yourself from all circles in the org, and you should still be able to see each circle's members page in the limited way described above.

If you want to test this in the preview app, you need to also deploy this code to staging, and then run the REPL against staging (by setting NODE_HASURA_URL and HASURA_GRAPHQL_ADMIN_SECRET to staging values)

### Todo

- [x] stop using useSelectedCircle
- [x] combine queries on members page
- [x] set up permissions so org members can view circle, users, profiles
- [x] move MemberRow to its own file
- [x] select correct org in side nav when viewing a circle you're not part of
- [x] show non-member circles in side nav only when selected
- [x] fix circle overview page (deferred to future PR)
- [x] get map page working (deferred to future PR)
- [x] debug failing assertion when switching from non-member page to give page for a circle you're in
  - root cause: useSelectedCircle called before it's updated for the new circle

levity commented:
Current dependencies on/for this PR:
* main
  * **PR #1993** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1993" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #1994** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1994" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #1995** <a href="https://app.graphite.dev/github/pr/coordinape/coordinape/1995" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/coordinape/coordinape/1995?utm_source=stack-comment).

levity commented on src/hooks/migration.ts:
```diff
@@ -27,20 +27,27 @@ export const useFixCircleState = (circleId: number | undefined) => {
     } else {
       logger.log(`fetching circle data: ${circleId}`);
-      fetchCircle({ circleId, select: true });
+      fetchCircle({ circleId, select: true }).then(() => setReady(true));
```
this prevents a race condition where a new circle's page starts rendering before the data for it is ready, i.e. it uses the previous circle's data

crabsinger replied:
Not sure this really would work with all this recoil stuff, but RQ exposes an isFetching prop that maybe could serve as a similar purpose to hold off on page load until the query is complete?

levity commented on src/hooks/useConnectedAddress.ts:
```diff
@@ -1,8 +1,11 @@
 import { useWeb3React } from 'hooks/useWeb3React';
 
-export default function useConnectedAddress() {
+function useConnectedAddress(required: true): string;
```
https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads

levity commented on src/recoilState/app.ts:
```diff
@@ -123,23 +123,13 @@ export const rCircle = selectorFamily<ICircleState, number | undefined>({
         circle,
         myUser: me,
         users: getCircleUsers().toArray(),
-        usersNotMe: getCircleUsers()
```
this was already unused

levity commented on src/pages/MembersPage/MembersTable.tsx:
```diff
@@ -931,38 +933,45 @@ export const MembersTable = ({
           />
         )}
       </MemberTable>
+      <LeaveCircleModal
```
moving this modal from MembersPage/index wasn't strictly necessary for this work; it just seemed opportune to refactor to reduce the amount of prop-drilling

levity commented on src/pages/MembersPage/MembersTable.tsx:
```diff
@@ -187,29 +184,28 @@ const UserName = ({ user }: { user: ICircleUser }) => {
 };
 
-const MemberRow = ({
+export const MemberRow = ({
```
i originally pulled this out into its own file but it made the diff much bigger

crabsinger commented:
Adding a note to this PR in case you didn't see this: https://github.com/coordinape/coordinape/pull/1944/files#r1131506108

crabsinger reviewed (commented):
UI:

I added my user to the org, but not the circle: I see the Activity, Map, and Members page show on the nav bar, but Activity and Map are not yet supported so they just redirect to /circles.

Feels a bit weird to ship with those links broken, maybe they're disabled or removed for now until those pages support org members viewing?
<img width="896" alt="Screenshot 2023-03-09 at 11 42 20 AM" src="https://user-images.githubusercontent.com/83605543/224136402-620fb858-0905-4f96-a790-f0c85fef3328.png">

crabsinger commented on src/features/nav/getNavData.ts:
```diff
@@ -15,14 +15,7 @@ export const getNavData = (address: string, chainId: number) =>
-                },
-              },
-            },
+            {},
```
maybe the default permissions handle this already for us?

levity replied:
yep, soft-deleted circles are excluded. this did prompt me to notice that we don't do the same for _users_ of soft-deleted circles tho; spun that out into #2023

levity replied:
i'll also make this change for the org_members table itself

crabsinger commented on src/pages/MembersPage/getMembersPageData.ts:
```diff
@@ -0,0 +1,110 @@
+
+import { Awaited } from 'types/shim';
+
+export const getMembersPageData = async (circleId: number) => {
```
maybe we should rename this to `CircleMembers` or `OrgMembers` now to differentiate these two...

levity replied:
I'll do this renaming in a later PR

crabsinger reviewed (commented):
a few comments and questions.

levity commented:
> UI:
> 
> 
> 
> I added my user to the org, but not the circle: I see the Activity, Map, and Members page show on the nav bar, but Activity and Map are not yet supported so they just redirect to /circles.
> 
> 
> 
> Feels a bit weird to ship with those links broken, maybe they're disabled or removed for now until those pages support org members viewing?
> 
> <img width="896" alt="Screenshot 2023-03-09 at 11 42 20 AM" src="https://user-images.githubusercontent.com/83605543/224136402-620fb858-0905-4f96-a790-f0c85fef3328.png">
> 
> 

Hmm, the map link showing is a bug, I thought I removed that. I didn't check activity because I don't have it enabled locally

crabsinger commented on hasura/metadata/databases/default/tables/public_users.yaml:
```diff
@@ -148,13 +148,25 @@ select_permissions:
+              - circle:
+                  users:
+                    _and:
+                      - deleted_at:
```
a bit pedantic,but I've been putting the `deleted_at` clause as second after the profile check consistently for easier scanning and readability

crabsinger reviewed (dismissed):
looks good!

levity commented:
@levity queued this pull request to merge with [Graphite](https://app.graphite.dev/github/pr/coordinape/coordinape/1995).

levity commented:
@levity merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coordinape/coordinape/1995).
